File: graph.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <string>
#include <iomanip>

using namespace std;

class Node {
public:
    int nodeNumber;
    double demand;
    double supply;
    
    Node() : nodeNumber(0), demand(0.0), supply(0.0) {}
    
    Node(int num, double dem = 0.0, double sup = 0.0) 
        : nodeNumber(num), demand(dem), supply(sup) {}
    
    void display() const {
        cout << "Node " << nodeNumber 
             << ": Demand=" << demand 
             << ", Supply=" << supply << endl;
    }
};

class PowerGridGraph {
private:
    map<int, Node> nodes;
    map<int, vector<pair<int, double>>> adjacencyList; // node -> [(neighbor, length)]
    
public:
    void addNode(const Node& node) {
        nodes[node.nodeNumber] = node;
    }
    
    void addEdge(int node1, int node2, double length) {
        // Ensure both nodes exist
        if (nodes.find(node1) == nodes.end()) {
            nodes[node1] = Node(node1);
        }
        if (nodes.find(node2) == nodes.end()) {
            nodes[node2] = Node(node2);
        }
        
        // Add directed edge from node1 to node2
        adjacencyList[node1].push_back(make_pair(node2, length));
    }
    
    bool loadDemandData(const string& filename) {
        ifstream file(filename);
        if (!file.is_open()) {
            cerr << "Error: Could not open file " << filename << endl;
            return false;
        }
        
        string line;
        getline(file, line); // Skip header
        
        while (getline(file, line)) {
            stringstream ss(line);
            string nodeStr, demandStr;
            
            getline(ss, nodeStr, ',');
            getline(ss, demandStr, ',');
            
            int nodeNum = stoi(nodeStr);
            double demand = -stod(demandStr); // Store demand as negative
            
            if (nodes.find(nodeNum) != nodes.end()) {
                nodes[nodeNum].demand = demand;
            } else {
                Node node(nodeNum, demand, 0.0);
                addNode(node);
            }
        }
        
        file.close();
        cout << "Loaded demand data from " << filename << endl;
        return true;
    }
    
    bool loadSupplyData(const string& filename) {
        ifstream file(filename);
        if (!file.is_open()) {
            cerr << "Error: Could not open file " << filename << endl;
            return false;
        }
        
        string line;
        getline(file, line); // Skip header
        
        while (getline(file, line)) {
            stringstream ss(line);
            string nodeStr, supplyStr;
            
            getline(ss, nodeStr, ',');
            getline(ss, supplyStr, ',');
            
            int nodeNum = stoi(nodeStr);
            double supply = stod(supplyStr);
            
            if (nodes.find(nodeNum) != nodes.end()) {
                nodes[nodeNum].supply = supply;
            } else {
                Node node(nodeNum, 0.0, supply);
                addNode(node);
            }
        }
        
        file.close();
        cout << "Loaded supply data from " << filename << endl;
        return true;
    }
    
    bool loadLinesData(const string& filename) {
        ifstream file(filename);
        if (!file.is_open()) {
            cerr << "Error: Could not open file " << filename << endl;
            return false;
        }
        
        string line;
        getline(file, line); // Skip header
        
        while (getline(file, line)) {
            stringstream ss(line);
            string lineNumStr, node1Str, node2Str, lengthStr;
            
            getline(ss, lineNumStr, ',');
            getline(ss, node1Str, ',');
            getline(ss, node2Str, ',');
            getline(ss, lengthStr, ',');
            
            int node1 = stoi(node1Str);
            int node2 = stoi(node2Str);
            double length = stod(lengthStr);
            
            addEdge(node1, node2, length);
        }
        
        file.close();
        cout << "Loaded lines data from " << filename << endl;
        return true;
    }
    
    void displayGraph() const {
        cout << "\nPOWER GRID GRAPH\n";
        cout << "\nNodes:\n";
        cout << "--------------------------------------\n";
        for (const auto& pair : nodes) {
            pair.second.display();
        }
        
        cout << "\nAdjacency List (Connections):\n";
        cout << "--------------------------------------\n";
        for (const auto& pair : adjacencyList) {
            cout << "Node " << pair.first << " connects to: ";
            for (const auto& edge : pair.second) {
                cout << "[Node " << edge.first << ", Length=" << edge.second << "] ";
            }
            cout << endl;
        }
    }
    
    int getNodeCount() const {
        return nodes.size();
    }
    
    int getEdgeCount() const {
        int count = 0;
        for (const auto& pair : adjacencyList) {
            count += pair.second.size();
        }
        return count; // Each directed edge is counted once
    }
    
    Node* getNode(int nodeNumber) {
        if (nodes.find(nodeNumber) != nodes.end()) {
            return &nodes[nodeNumber];
        }
        return nullptr;
    }
    
    vector<pair<int, double>> getNeighbors(int nodeNumber) {
        if (adjacencyList.find(nodeNumber) != adjacencyList.end()) {
            return adjacencyList[nodeNumber];
        }
        return vector<pair<int, double>>();
    }
};

int main() {
    PowerGridGraph graph;
    
    cout << "Building Power Grid Graph...\n" << endl;
    
    // Load the three CSV files
    graph.loadDemandData("dataset/Gen_WI_Demand_Values.csv");
    graph.loadSupplyData("dataset/Gen_WI_Supply_Values.csv");
    graph.loadLinesData("dataset/Gen_WI_Lines.csv");
    
    // Display the graph
    graph.displayGraph();
    
    // Display summary statistics
    cout << "\n========== GRAPH STATISTICS ==========\n";
    cout << "Total Nodes: " << graph.getNodeCount() << endl;
    cout << "Total Edges: " << graph.getEdgeCount() << endl;
    
    // Example: Access a specific node
    cout << "\n========== EXAMPLE USAGE ==========\n";
    Node* node = graph.getNode(12075);
    if (node != nullptr) {
        cout << "Accessing Node 12075:\n";
        node->display();
        
        cout << "\nNeighbors of Node 12075:\n";
        vector<pair<int, double>> neighbors = graph.getNeighbors(12075);
        for (const auto& neighbor : neighbors) {
            cout << "  -> Node " << neighbor.first 
                 << " (Distance: " << neighbor.second << ")\n";
        }
    }
    
    return 0;
}

File: graph.py
import csv

class Node:
    def __init__(self, node_number, demand=0.0, supply=0.0):
        self.node_number = node_number
        self.demand = demand
        self.supply = supply

    def display(self):
        print(f"Node {self.node_number}: Demand={self.demand}, Supply={self.supply}")


class PowerGridGraph:
    def __init__(self):
        self.nodes = {}  # node_number -> Node object
        self.adjacency_list = {}  # node_number -> [(neighbor, length)]

    def add_node(self, node):
        self.nodes[node.node_number] = node

    def add_edge(self, node1, node2, length):
        # Ensure nodes exist
        if node1 not in self.nodes:
            self.add_node(Node(node1))
        if node2 not in self.nodes:
            self.add_node(Node(node2))

        # Add directed edge
        self.adjacency_list.setdefault(node1, []).append((node2, length))

    # -----------------------------
    # Load demand CSV
    # -----------------------------
    def load_demand_data(self, filename):
        try:
            with open(filename, "r") as file:
                reader = csv.reader(file)
                next(reader)  # skip header

                for row in reader:
                    node_num = int(row[0])
                    demand = -float(row[1])  # negative demand

                    if node_num in self.nodes:
                        self.nodes[node_num].demand = demand
                    else:
                        self.add_node(Node(node_num, demand=demand))

            print(f"Loaded demand data from {filename}")
            return True

        except FileNotFoundError:
            print(f"Error: Could not open file {filename}")
            return False

    # -----------------------------
    # Load supply CSV
    # -----------------------------
    def load_supply_data(self, filename):
        try:
            with open(filename, "r") as file:
                reader = csv.reader(file)
                next(reader)

                for row in reader:
                    node_num = int(row[0])
                    supply = float(row[1])

                    if node_num in self.nodes:
                        self.nodes[node_num].supply = supply
                    else:
                        self.add_node(Node(node_num, supply=supply))

            print(f"Loaded supply data from {filename}")
            return True

        except FileNotFoundError:
            print(f"Error: Could not open file {filename}")
            return False

    # -----------------------------
    # Load lines CSV
    # -----------------------------
    def load_lines_data(self, filename):
        try:
            with open(filename, "r") as file:
                reader = csv.reader(file)
                next(reader)

                for row in reader:
                    node1 = int(row[1])
                    node2 = int(row[2])
                    length = float(row[3])
                    self.add_edge(node1, node2, length)

            print(f"Loaded lines data from {filename}")
            return True

        except FileNotFoundError:
            print(f"Error: Could not open file {filename}")
            return False

    # -----------------------------
    def display_graph(self):
        print("\nPOWER GRID GRAPH\n")
        print("Nodes:")
        print("--------------------------------------")

        for node in self.nodes.values():
            node.display()

        print("\nAdjacency List (Connections):")
        print("--------------------------------------")

        for node, edges in self.adjacency_list.items():
            edge_info = ", ".join([f"(Node {n}, Length={l})" for n, l in edges])
            print(f"Node {node} connects to: {edge_info}")

    # -----------------------------
    def get_node_count(self):
        return len(self.nodes)

    def get_edge_count(self):
        return sum(len(edges) for edges in self.adjacency_list.values())

    def get_node(self, node_number):
        return self.nodes.get(node_number, None)

    def get_neighbors(self, node_number):
        return self.adjacency_list.get(node_number, [])


# ============================================================
# Main Equivalent
# ============================================================

if __name__ == "__main__":
    graph = PowerGridGraph()

    print("Building Power Grid Graph...\n")

    graph.load_demand_data("dataset/Gen_WI_Demand_Values.csv")
    graph.load_supply_data("dataset/Gen_WI_Supply_Values.csv")
    graph.load_lines_data("dataset/Gen_WI_Lines.csv")

    # Display graph
    graph.display_graph()

    # Stats
    print("\n========== GRAPH STATISTICS ==========")
    print("Total Nodes:", graph.get_node_count())
    print("Total Edges:", graph.get_edge_count())

    # Example usage
    print("\n========== EXAMPLE USAGE ==========")
    node = graph.get_node(12075)
    if node:
        print("Accessing Node 12075:")
        node.display()

        print("\nNeighbors of Node 12075:")
        neighbors = graph.get_neighbors(12075)
        for n, dist in neighbors:
            print(f"  -> Node {n} (Distance: {dist})")
